week_1/Array/longest_substring_without_repeating_characters/longest_substring_without_repeating_characters.js

```diff
@@ -0,0 +1,33 @@
+//Given a string s, find the length of the longest substring without repeating characters.
+/**
+ * @param {string} s
+ * @return {number}
+ */
+const lengthOfLongestSubstring = (s) => {
+	if (s.length === 0 || s.length === 1) return s.length;
+
+	let start = 0;
+	let longest = 1;
+	let lastSeen = new Map();
+
+	for (let end = 0; end < s.length; end++) {
+		let currentLetter = s[end];
+
+		// If the current letter has previously been seen, move the start index up to one letter ahead of the last seen index
+		if (lastSeen.has(currentLetter) && lastSeen.get(currentLetter) >= start) {
```

Comment:
Performance nitpick: calling `has` and `get` right next to each other requires a call into the map twice. We can rely on `get`'s behavior of returning `undefined` if the key doesn't exist to improve our performance a bit.

```js
const seen = lastSeen.get(currentLetter);
if (seen && seen >= start)
```

Reply:
This is a good watch related to this 😄 From the creator of c++ https://www.youtube.com/watch?v=YQs6IC-vgmo